k2/csrc/pytorch_context.cu

```diff
@@ -116,8 +117,21 @@ class PytorchCudaContext : public Context {
 ContextPtr GetCpuContext() { return std::make_shared<PytorchCpuContext>(); }
 
 ContextPtr GetCudaContext(int32_t gpu_id /*= -1*/) {
-  if (gpu_id < 0) gpu_id = c10::cuda::current_device();
-  return std::make_shared<PytorchCudaContext>(gpu_id);
+  static std::once_flag has_cuda_init_flag;
```

Comment:
how about we put those `call_once` code in to a function and return `has_cuda`, then there will be no duplicate code, another advantage is we really print only once the warning if we call this in more than 1 function (e.g. context.cu and time.cu

Reply:
I think timer.cu will not need it since Timer will accept a ContextPtr,  which cannot be a cuda context when GetCudaContext() returns a cpu context.

Reply:
Yeah, I just take the current code as an example to show that we will print warning twice if we do the same thing in different functions. Anyway, let's remove the duplicate code if possible.

Reply:
Fixed.

Reply:
RE the naming, since it's kind of an `Impl` paradigm, I wonder whether it would make sense to call BaseTimer TimerImpl, and have CpuTimerImpl and CudaTimerImpl.  I have been avoiding the word `Gpu` in these names because one day we may want to support AMD Gpu's or other GPUs.

Reply:
Fixed. Renamed to TimerImpl.